### PR TITLE
UI Modifications and some bug fixes

### PR DESCRIPTION
**Pull request by sp4ce-cowboy**
### Overview

> Previously, each transaction would be appended to the bottom of the transaction list panel, which would impede the user from getting immediate feedback from their input, as the TransactionListPanel does not scroll automatically.

> Rather than making the panel auto-scroll, the order of the list is reversed so that the user is always looking at the most
recent transaction, which means that the user would not have to move away from the keyboard to get a response. The index associated with each TransactionCard remains the same (i.e. reversed together in the same way), as before, i.e. there would be no change in logic required for commands that require transaction index, WYSIWYG.

> Additionally, CommandBox has been modified to span the whole Window with a smaller font size so that it can accomodate a longer input command. This would also mean that the CommandBox doesn't need to expand to occupy 2 lines as previously discussed.

Continued from this [PR](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/94), some more minor modifications have been made, including the generation of category colors from category hash and other miscellaneous bug fixes.

Screenshot 2023-10-23 at 12 33 17 PM

### Changes Made:
- Ui Changes required for the above, including StyleSheet modifications
- Added a rolling balance indicator within the StatusBarFooter
- Added a proper UniCash Welcome Message that collates all availble `COMMAND_WORDS`
- Changes to the Amount class and the handling of currency symbols, validation of numbers
    - Note that for the above, tests have yet to be implemented as @lipwei1808 will continue to work on them
- Removal of case-sensitivity for categories and improved Duplicates detection
- Fixed some other very minor bugs I noticed, including one in `EditTransactionsCommand`

### Some future considerations:
- Addition of a LocalDateTime.now() within every Transaction to associate a creation time
- Which can then be used for sorting (Sort by transaction dateTime or transaction creation Time)
- Addition of support for the user to enter `$`-prefixed amounts

### CODE REVIEW

File: src/main/java/unicash/model/category/Category.java
Comment (sp4ce-cowboy) [inline src/main/java/unicash/model/category/Category.java:29]: If we wanted to ever revert to case-sensitive categorizing later, all we would need to do is just remove this part

File: src/test/java/unicash/model/category/UniqueCategoryListTest.java
Comment (sp4ce-cowboy) [inline src/test/java/unicash/model/category/UniqueCategoryListTest.java:76]: this line was modified because the case-insensitivity of the categories were causing `test4` to be flagged as a duplicate

File: src/main/java/unicash/model/category/Category.java
Comment (lipwei1808) [reply src/main/java/unicash/model/category/Category.java:29]: Could leave a comment in the Category java docs that mentions all Category's are case insensitive?

File: src/main/java/unicash/model/category/Category.java
Comment (lipwei1808) [inline src/main/java/unicash/model/category/Category.java:69]: Small nit, the naming of this method might be unintuitive. I almost intepreted it as get the hash value of the tag and return the string

File: src/main/java/unicash/ui/TransactionCard.java
Comment (lipwei1808) [inline src/main/java/unicash/ui/TransactionCard.java:72]: Is there a reason to hardcode each category to a label instead of defining them inside something like a FlowPane?

File: src/main/java/unicash/model/commons/Amount.java
Comment (lipwei1808) [inline src/main/java/unicash/model/commons/Amount.java:24]: Do we want to immediately truncate all Amount values as we store them? An alternative would be to still store as the value inputted but only alter the UI to display 2dp. I think that might be better instead of modifying what the user inputted

Comment (lipwei1808) [reply src/main/java/unicash/model/commons/Amount.java:24]: Else, could add some documentation in the class to state that the Amount only stores 2dp values and automatically truncates

File: src/main/java/unicash/model/commons/Amount.java
Comment (lipwei1808) [inline src/main/java/unicash/model/commons/Amount.java:67]: Is this method still used anywhere, if not maybe we could remove it

File: src/main/java/unicash/ui/StyleSheet.java
Comment (lipwei1808) [inline src/main/java/unicash/ui/StyleSheet.java:52]: Is it possible to not hardcode this? Since this is assuming the hexString is always at least 5 charcaters long right. Like some padding function by String

**Review by lipwei1808 — COMMENTED** [review]
Great work, just some questions on some decisions you made. Nice StyleSheet.java to facilitate the Category background colors. Maybe could set the text colour to white, else it's abit hard to read

File: src/main/java/unicash/model/commons/Amount.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/commons/Amount.java:24]: I'll reply to this before the other ones cus it's important - the reason is, if we only change the values at the UI level, there are a lot of discrepancies that arise and those are only just the ones I noticed yesterday.

For example, editing a transaction with a non-2dp amount would reflect the 2dp value in the TransactionCard, but the ResultDisplay box would still show the exact value, and that's because that information comes from UniCashMessage's formatTransaction method. Another example would be budgetting or balance calculating - any testers/users might have issues when their balance or budget is calculated when fringe amounts or too precise amounts are entered. e.g. A budget set at $200 might never be internally mathematically flagged as reached if the user inputs 199.999999 but any UI components depending on that will be modified and flagged to have been exceeded.

Therefore, I strongly suggest that we truncate all amounts to a maximum of 2 decimal places from the very point of instantiation so that no other discrepancies like this can arise later on, and we can completely avoid any possible bugs. Additionally, I think it also suits our target user profile as University students won't be dealing with any transactions that would require a higher than standard currency precision (like trading/currency exchange I guess? I'm actually not too sure abt transactions more precise than 2dp irl)

File: src/main/java/unicash/model/commons/Amount.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/commons/Amount.java:67]: It's for backwards compatibility with some failing unit tests, one of which requires some toString to be parsed, but I can go back and check if I can change those tests

File: src/main/java/unicash/ui/TransactionCard.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/TransactionCard.java:72]: Good idea, I actually didn't think of that I just went with the more obvious way.

File: src/main/java/unicash/model/category/Category.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/category/Category.java:69]: you're right I'll get that fixed

File: src/main/java/unicash/model/commons/Amount.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/commons/Amount.java:67]: `parseCommand_addTransaction` in UniCashParserTest seems to require it, I think it's supp to parse simulate user inputs, and since the dollar symbol prefixed is not a recognized user input for amount, the regular `toString` method doesn't work. Thinking about, maybe we can allow the user to input a currency symbol as well? like `amt/$40.20` seems like a natural input right

File: src/main/java/unicash/model/commons/Amount.java
Comment (lipwei1808) [reply src/main/java/unicash/model/commons/Amount.java:35]: Sure, I think that sounds good, just have to add in some proper documentation of this in the class and our DG.

File: src/main/java/unicash/ui/TransactionCard.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/TransactionCard.java]: Ok after trying FlowPane, the styling changed q a bit and I'm not sure how to fix it, but in the mean time as I work on fixing that, and since the labels being tied to each category doesn't affect functionality nor the UI directly, I think that we can move on with just the Labels for now, and at a later iteration we can think about simplifying the code, after the more important major features and functionalities have been implemented.

**Comment by sp4ce-cowboy** [comment]
The failing Codecov is due to the UI (which is to be expected) but also changes to the Amount class. The changes to the Amount class are not yet functional, I'm adding it together with this PR so that @lipwei1808 can pick up from here and work on both making the changes functional and adding the required unit tests.

File: src/main/java/unicash/logic/UniCashMessages.java
Comment (woojiahao) [inline src/main/java/unicash/logic/UniCashMessages.java:49]: Any reason why the newlines are inconsistent? Also, I'm thinking we should really have delegating this job to StringBuilder, this is very hard to read

File: src/main/java/unicash/logic/commands/EditTransactionCommand.java
Comment (woojiahao) [inline src/main/java/unicash/logic/commands/EditTransactionCommand.java:37]: This should really be a StringBuilder or a custom class to handle our success messages, this pattern is way too prone to human error

Comment (woojiahao) [reply src/main/java/unicash/logic/commands/EditTransactionCommand.java:37]: Something like having a `UsageBuilder` that takes in `Option... options` and `String usage` parameters where `Option` has `prefix`, `name`, and `isOptional` fields, this way, we can easily compose the resulting string without needing to resort to manually writing out these lengthy String concatenations for all commands, including the new ones

File: src/main/java/unicash/model/commons/Amount.java
Comment (woojiahao) [inline src/main/java/unicash/model/commons/Amount.java:13]: Was wondering, but why is duplicate work to add constraints done in this PR? I believe #97 is also handling the constraints, is there a reason why we duplicated the constraint handling? 

Only asking since UI modifications does not seem to entail including constraint checking. 

No issues here but maybe we can keep PRs to be local to the problems we're addressing and avoid adding changes outside of the scope to avoid potential merge conflicts/duplicate work done, this way it's also easier to know at a glance what the PRs are addressing

File: src/main/java/unicash/model/commons/Amount.java
Comment (woojiahao) [inline src/main/java/unicash/model/commons/Amount.java:86]: These checks don't really make sense to me, would the "range" of our amount already be `>= 0.00`? `isWithinRange` behavior does not line up with the method behavior too since it's supposed to be "non-negative" but you're checking for $-2^{31}$

Might I suggest collating these checks into a single `isWithinRange` that properly enforces the range of $[0.00, 2^{31} - 1]$

File: src/main/java/unicash/model/commons/Amount.java
Comment (woojiahao) [inline src/main/java/unicash/model/commons/Amount.java:101]: Maybe instead of trying to do string manipulations this way, we can write a quick Regex to handle the checking instead? Much like how `Location` and `Category` does their checking against a String input? This way, it's easier to understand the behavior at a glance and we could expand the support decimal places without much issue in the future

File: src/main/java/unicash/ui/StatusBarFooter.java
Comment (woojiahao) [inline src/main/java/unicash/ui/StatusBarFooter.java:25]: Maybe make this a global string inside our `utils` package. This way if we do change the decimal formatting, we don't have to edit different places and potentially miss one out

File: src/main/java/unicash/ui/StatusBarFooter.java
Comment (woojiahao) [inline src/main/java/unicash/ui/StatusBarFooter.java:44]: Can omit the use of `this`

File: src/main/java/unicash/ui/StatusBarFooter.java
Comment (woojiahao) [inline src/main/java/unicash/ui/StatusBarFooter.java:58]: Can omit use of `this`

File: src/main/java/unicash/ui/StatusBarFooter.java
Comment (woojiahao) [inline src/main/java/unicash/ui/StatusBarFooter.java:84]: Could we use the pre-assigned `TransactionType.EXPENSE.type` String here instead? This way if we do rename the enum string, we don't leave this to have undefined behavior

File: src/main/java/unicash/ui/StyleSheet.java
Comment (woojiahao) [inline src/main/java/unicash/ui/StyleSheet.java:74]: If we're using this in the constraints of a `String`, can we make the type parameter a `String` instead? I don't personally see a compelling reason for `Object` to be used since it leaves potentially "undesired" behavior as we perform unit testing/during PE testing, like having to account for different input types

File: src/main/java/unicash/ui/TransactionCard.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionCard.java:141]: Can omit the use of `this`

File: src/main/java/unicash/ui/TransactionCard.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionCard.java:156]: Can we compare the type by using the actual enumeration? Similar reason to a comment made earlier about providing a single source of truth

File: src/main/java/unicash/ui/TransactionCard.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionCard.java:185]: Is this a constant in `Stylesheet`? If it's not yet, this makes aa great candidate for a constant so there isn't dangling changes

File: src/main/java/unicash/ui/TransactionCard.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionCard.java:203]: Could the declaration be 

```java
List<Label> labelArrayList = Arrays.asList(firstCategory, secondCategory, thirdCategory, fourthCategory, fifthCategory)
``` 

instead? Since it doesn't seem like  you're modifying this array list so you can use a `List` instead of `ArrayList`

File: src/main/java/unicash/ui/TransactionCard.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionCard.java:214]: If the only use for the `labelsArrayList` is to call `.get(i)`, maybe it's better served to use a fixed size array instead, not a major issue but just good convention

File: src/main/java/unicash/ui/TransactionListPanel.java
Comment (woojiahao) [inline src/main/java/unicash/ui/TransactionListPanel.java:68]: Can omit the use of `this`

File: src/test/java/unicash/testutil/TransactionUtil.java
Comment (woojiahao) [inline src/test/java/unicash/testutil/TransactionUtil.java:37]: Just curious but is this `originalString` going to be cached when we store to storage? If it is, no issues, but if it's not, this might be an issue in terms of "preserving" the value

File: src/test/java/unicash/ui/StyleSheetTest.java
Comment (woojiahao) [inline src/test/java/unicash/ui/StyleSheetTest.java:31]: Not sure if we do need to test the constants. These tests are quite inflexible as they don't test every constant and if edited to do so, it'll be a pain to maintain if we change the constants since we basically need to duplicate the values

File: src/test/java/unicash/ui/StyleSheetTest.java
Comment (woojiahao) [inline src/test/java/unicash/ui/StyleSheetTest.java:68]: If the reason for making the input an `Object` for testing purposes, I strongly believe constraining the inputs to `Category` or `String` and then properly generating those values to generate the expected hex codes should be more appropriate

**Review by woojiahao — CHANGES_REQUESTED** [review]
Great work! @sp4ce-cowboy Although there are some consistency/duplicate work/questions I have about code design choices so hope to get your insights about them

File: src/main/java/unicash/model/commons/Amount.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/commons/Amount.java:13]: Yes actually this PR was not supposed to deal with the Amount class at all, but due to some formatting discrepanices of Amounts (also mentioned in [this](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/96#discussion_r1367847714) comment) that were problematic at the UI level it became the case. While discussing with @lipwei1808 I had some suggestions to the Amount class that I added together with this PR (which is why also there are no Amount Tests yet) so that he can take over as I've informed him [prior](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/97#discussion_r1368146067). I guess as opposed to finding the stash and copying over the .java file to send it manually, this seemed a bit more expedient. Either way, I can revert to the previous Amount class if this arrangement might cause any issues.

File: src/main/java/unicash/ui/StatusBarFooter.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/StatusBarFooter.java:25]: Thinking about it, it actually doesn't make sense for the balance formatter to be here at all. I've actually gone ahead and removed this entirely, and `StatusBarFooter` now relies on `Amount`'s String representation alone. I think this makes more sense as `Amount` has its own currency symbol and precision formatting, any indicator or label pertaining to amounts should not override whatever is specified by `Amount`.

File: src/main/java/unicash/ui/StyleSheet.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/StyleSheet.java:74]: This occured to me while speaking with @elhy1999 about his Statistics representation, I realized that this process of getting a color from an `Object` can be applied universally across the application, whereever there is a UI element with colors involved, this can be used to generate a color specific to any element. Given that this method also has built in color correction processes, all UI elements (or really, anything with a hash code, which is everything) involving color would have a consistent color scheme. 

Ngl I honestly didn't really think of this when writing the method initially but now I think it's a pretty neat way to ensure color consistency across our App.

File: src/main/java/unicash/ui/TransactionCard.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/TransactionCard.java]: Actually this part was a adhoc UI test that I missed out, I've removed those lines, thanks for noticing!

File: src/test/java/unicash/testutil/TransactionUtil.java
Comment (sp4ce-cowboy) [reply src/test/java/unicash/testutil/TransactionUtil.java:37]: I'm not entirely sure what you mean, but this change was also discussed [here](https://github.com/AY2324S1-CS2103-T16-3/tp/pull/96#discussion_r1367844482). If you are referring to the way Transactions are stored, in `JsonAdaptedTransaction` I noticed that an Amount's `amount` field is retrieved directly, maybe you could further clarify what you mean?

File: src/main/java/unicash/model/commons/Amount.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/model/commons/Amount.java:86]: I've edited this method to make more sense, there was an error in the documentation, thanks for noticing. 

About the collation of these checks and the below-mentioned changes to the Amount class, I suggest I let @lipwei1808 work on them instead? So that there won't be any duplication of work.

File: src/main/java/unicash/logic/commands/EditTransactionCommand.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/logic/commands/EditTransactionCommand.java:37]: As for this and the above comment, may I suggest that @jamz903 work on them alongside the `COMMAND_WORD` modifcations in v1.3b? I think it would be quite appropriate to have both `COMMAND_WORD` and `MESSAGE_USAGE` messages unified and accessible from the same place. That way, we would have a centralized location for defining/modifying user inputs and `ResultDisplay` outputs, perhaps we could also unify that with the existing `CliSyntax` as well, since everywhere that `MESSAGE_USAGE` exists, `CliSyntax.PREFIX` seems to exist as well, what do you think?

File: src/main/java/unicash/ui/StyleSheet.java
Comment (woojiahao) [reply src/main/java/unicash/ui/StyleSheet.java]: If that is the case, then could we consider using using generics and method overloading to constraint the method signatures? I agree it's a neat way to ensure color consistency but I am against leaving the parameters as just a generic `Object`. 

Since we're applying this to any UI elements, it could be a generic that extends the `UiPart` type or `Category` type and both lead to a common method that takes in an `int hash` instead of a pure `Object`. 

From the perspective of testing, an `Object` parameters requires way more extensive testing as it can be anything that goes in, this comment is more so to raise the potential code design flaw of just working with pure Objects when we can perhaps constraint it further using generics and avoid any foot guns with using Objects.

File: src/test/java/unicash/testutil/TransactionUtil.java
Comment (woojiahao) [reply src/test/java/unicash/testutil/TransactionUtil.java:37]: Gotcha, for some reason `originalString` gives off the impression that the function returns either `$xx.xx` or `xx.xx` since original could refer to the original format that the amount came in.

If that's the case, maybe we can focus on making the method name more obvious with the purpose like `amountString()` or something similar so the code reads more smoothly?

**Review by woojiahao — CHANGES_REQUESTED** [review]
Most of the changes looks good! Would still need to discuss the use of `Object` before we merge this PR as I think it's a potential thing we could be graded down on if not careful

File: src/main/java/unicash/ui/StyleSheet.java
Comment (sp4ce-cowboy) [reply src/main/java/unicash/ui/StyleSheet.java]: Sure, that makes sense, I'll commit the changes and modify the StyleSheet unit tests as well.

**Comment by sp4ce-cowboy** [comment]
> Most of the changes looks good! Would still need to discuss the use of `Object` before we merge this PR as I think it's a potential thing we could be graded down on if not careful

Thanks @woojiahao for the detailed comments, I appreciate the effort you've taken to throughly review the code. I've incorporated all the changes you've requested, if there's anyth else I missed out do let me know.

**Review by woojiahao — CHANGES_REQUESTED** [review]
LGTM! There's one more open conversation that's left about using Regex instead of using manual string parsing for checking if an amount has the right number of decimal places

**Comment by sp4ce-cowboy** [comment]
> LGTM! There's one more open conversation that's left about using Regex instead of using manual string parsing for checking if an amount has the right number of decimal places

https://github.com/AY2324S1-CS2103-T16-3/tp/pull/96#discussion_r1368456979

File: src/main/java/unicash/model/commons/Amount.java
Comment (woojiahao) [reply src/main/java/unicash/model/commons/Amount.java]: Ah, but if @lipwei1808 was already leaving out the `Amount` constraint checks for you to handle, wouldn't it make more sense to make the change here?

File: src/main/java/unicash/model/commons/Amount.java
Comment (woojiahao) [inline src/main/java/unicash/model/commons/Amount.java:160]:
```suggestion
        return String.format("%s%.2f", CURRENCY_INDICATOR, amount);
```

**Review by woojiahao — APPROVED** [review]
LGTM, just one suggestion to inline the `toString` with the use of `String.format`, don't need to do String concatenation if you're already using `String.format`

**Pull request merged**
